### DISCUSSION TITLE
Override linkage in manifest

### DISCUSSION THREAD

**Discussion by sandercox** [microsoft/vcpkg#19193] [Ideas]
I've been working with vcpkg in commercially distributed software. As such I need to watch the license of packages to determine how I can link them with our software. For example ffmpeg is LGPL so it needs to be linked as a shared library to not violate the L-GPL license.

Though I like to use as much as possible thru static libraries to allow compilers to optimize code in the entire program including 3rd party libraries and keep distribution a bit easier. Right now I'm using a custom vcpkg triplet or actuall an overlay of the default triplet names to 'fix' linkage based on port names.

This sort of works, but it requires a complete rebuild of all my dependencies everytime I'm adding a new entry to my manifest that doesn't conform to my default library linkage. This is annoying.

**Proposed solution**
It would be great if I could override for a port that it needs to be linked different.

**Describe alternatives you've considered**
Something else that could be interesting is to have information about the license in the CONTROL file of each port. With access to the license string then also in the triplet one could write something like:

```cmake
if(${LICENSE} MATCHES "LGPL")
  set(VCPKG_LIBRARY_LINKAGE dynamic)
endif()
```
But I assume that the CONTROL files are not parsed before reading the triplet config - didn't check that yet.

**Comment by JackBoosY** [comment]
Ping @strega-nil lest I forget something, we don't have this feature, right?

**Comment by dg0yt** [comment]
Configuration outside scripts, triplets and portfiles may create problems with binary caching. So I'm not sure if the following should be recommended: AFAIU only the triplet file content itself is part of the ABI tracking. What if you use an include statement to load specialized configuration, at least during development? You would still need to carefully track what is in the build at release time. 
A versioned name of the included file, or a comment in the tripet, may help to rebuild the world when needed.

(But license tracking is an issue which perhaps needs more attention. With a GPL app, my desire is to keep out features which are not compatible with GPL. I think will need some override ports with a 'GPL ' feature'. 'LGPL' might be an alternative feature.)

**Comment by sandercox** [comment]
It's been a while when I was working with it but I don't think you can include files from the triplet files. Because I wanted these changes for all our platforms I was trying to combine the checks into a single file and include that from the triplet files.

Didn't get that to work back then but things might have changed since.

**Comment by omartijn** [comment]
Nowadays, the license is part of `vcpkg.json`. Not all parts seems to have it yet, but that's usually trivially fixed, with a simple (both to create and review) PR. If you add the following to your triplet files, you should get a `PORT_LICENSE` variable that you can use:

```
set(PORT_LICENSE)

if (EXISTS "${CURRENT_PORT_DIR}/vcpkg.json")
    file(READ "${CURRENT_PORT_DIR}/vcpkg.json" PORT_JSON)
    string(JSON PORT_LICENSE ERROR_VARIABLE IGNORE_ERROR GET ${PORT_JSON} license)
endif()
```

Of course, then it's up to you to determine which licenses would be safe to statically link. If no license is found, `PORT_LICENSE` is set to `NOT-FOUND`.
